OpenSim/Common/MultivariatePolynomialFunction.cpp

```diff
@@ -17,205 +17,362 @@
+            const std::vector<std::string>& vars) {
+
+        OPENSIM_THROW_IF(static_cast<int>(vars.size()) != dimension, 
+                Exception,
+                "Expected the number of variable names ({}) to match the "
+                "polynomial dimension ({}), but it does not.", 
+                vars.size(), dimension)
+
+        int numCoefficients = choose(dimension + order, order);
+        std::vector<Term> terms;
+        terms.resize(numCoefficients);
+
+        std::vector<std::vector<int>> combinations;
+        generateCombinations(dimension, order, combinations);
+        for (int i = 0; i < numCoefficients; ++i) {
+            Term term;
+            for (int j = 0; j < dimension; ++j) {
+                if (combinations[i][j] > 0) {
+                    term[vars[j]] = combinations[i][j];
+                }
+            }
+            terms.at(i) = term;
+        }
+        
+        for (int i = 0; i < numCoefficients; ++i) {
+            (*this)[terms[i]] = i;
+        }
+    }
+
+    MultivariatePolynomial(
+            const MultivariatePolynomialFunction& mvp, 
+            const std::vector<std::string>& vars) {
+
+        int dimension = mvp.getDimension();
+        int order = mvp.getOrder();
+        SimTK::Vector coefficients = mvp.getCoefficients();
+
+        OPENSIM_THROW_IF(
+                coefficients.size() != choose(dimension + order, order), 
+                Exception,
+                "Expected the number of coefficients ({}) to match the number "
+                "of coefficients in a polynomial of dimension {} and order {}, "
+                "but it does not.", coefficients.size(), dimension, order);
+
+        std::vector<Term> terms;
+        terms.resize(coefficients.size());
+
+        OPENSIM_THROW_IF(static_cast<int>(vars.size()) != dimension, 
+                Exception,
+                "Expected the number of variable names ({}) to match the "
+                "polynomial dimension ({}), but it does not.", 
+                vars.size(), dimension)
+
+        std::vector<std::vector<int>> combinations;
+        generateCombinations(dimension, order, combinations);
+        for (int i = 0; i < coefficients.size(); ++i) {
+            Term term;
+            for (int j = 0; j < dimension; ++j) {
+                if (combinations[i][j] > 0) {
+                    term[vars[j]] = combinations[i][j];
+                }
+            }
+            terms.at(i) = term;
+        }
+        
+        for (int i = 0; i < coefficients.size(); ++i) {
+            (*this)[terms[i]] = coefficients[i];
+        }
+    }
+
+    // CALC METHODS
+    /**
+     * Calculate the order of the polynomial.
+     */
+    int calcOrder() const {
+        int order = 0;
+        for (auto it = begin(); it != end(); it++) {
+            int sum = 0;
+            for (auto it2 = it->first.begin(); it2 != it->first.end(); it2++) {
+                sum += it2->second;
```

Comment:
This loop, and the ones below it, could really use short comments to describe what is going on.